src/cpp/hu.bme.mit.inf.modes3.components.traindetector/traindetector_tests/tests/detection_test.cpp

```diff
@@ -0,0 +1,110 @@
+#include <gtest/gtest.h>
+#include <traindetector.h>
+
+TEST(DetectionTest, RegularLocomotive) {
+    TrainDetector detector;
+
+    detector.handleEdge(RIGHT, RISING); // 1
+    detector.handleEdge(LEFT, RISING); // 1
+    detector.handleEdge(RIGHT, FALLING); // 2
+    detector.handleEdge(LEFT, FALLING); // 2
+
+    auto detections = detector.getDetections();
+    ASSERT_EQ(detections.size(), 1);
+    ASSERT_EQ(detections[0].state, DetectionState::COMPL_STR);
+}
+
+TEST(DetectionTest, DoubleLocomotive) {
+    TrainDetector detector;
+
+    detector.handleEdge(RIGHT, RISING);  // 1
+    detector.handleEdge(LEFT, RISING);  // 1
+    detector.handleEdge(RIGHT, RISING);  // 2
+    detector.handleEdge(LEFT, RISING);  // 2
+    detector.handleEdge(RIGHT, FALLING); // 1
+    detector.handleEdge(LEFT, FALLING); // 1
+    detector.handleEdge(RIGHT, FALLING); // 2
+    detector.handleEdge(LEFT, FALLING); // 2
+
+    auto detections = detector.getDetections();
+    ASSERT_EQ(detections.size(), 2);
+    ASSERT_EQ(detections[0].state, DetectionState::COMPL_STR);
+    ASSERT_EQ(detections[1].state, DetectionState::COMPL_STR);
+}
+
+//TEST(DetectionTest, RevStrtLocomotive) {
+//    TrainDetector detector;
+//
+//    // Invalid train movement, moving in reverse after
+//    // covering both sensors
+//    // Should not detect, but detections after should
+//    // be valid
+//    detector.handleEdge(RIGHT, RISING);  // 1
+//    detector.handleEdge(LEFT, RISING);   // 1
+//    detector.handleEdge(LEFT, FALLING);  // 1
+//    detector.handleEdge(RIGHT, FALLING); // 1
+//    // Corrent train movement, should be detected
+//    detector.handleEdge(LEFT, RISING);   // 2
+//    detector.handleEdge(RIGHT, RISING);  // 2
+//    detector.handleEdge(LEFT, FALLING);  // 2
+//    detector.handleEdge(RIGHT, FALLING); // 2
+//
+//    auto detections = detector.getDetections();
+//    ASSERT_EQ(detections.size(), 2);
+//    ASSERT_EQ(detections[0].state, DetectionState::COMPL_STR);
+//    ASSERT_EQ(detections[1].state, DetectionState::COMPL_REV);
+//}
+
+TEST(DetectionTest, TrainWithTwoCars) {
+    TrainDetector detector;
+
+    detector.handleEdge(RIGHT, RISING);  // 1
+    detector.handleEdge(LEFT, RISING);   // 1
+    detector.handleEdge(RIGHT, FALLING); // 1
+    detector.handleEdge(RIGHT, RISING);  // 2
+    detector.handleEdge(LEFT, FALLING);  // 1
+    detector.handleEdge(LEFT, RISING);   // 2
+    detector.handleEdge(RIGHT, FALLING); // 2
+    detector.handleEdge(LEFT, FALLING);  // 2
+
+    auto detections = detector.getDetections();
+    ASSERT_EQ(detections.size(), 2);
+    ASSERT_EQ(detections[0].state, DetectionState::COMPL_STR);
+    ASSERT_EQ(detections[1].state, DetectionState::COMPL_STR);
+}
+
+
+TEST(DetectionTest, TrainWithThreeCars) {
```

Comment:
Separete the big test into smaller ones, so that after ASSERT there is no 'Act' anymore.